lib/std/compress/deflate.zig

```diff
@@ -481,32 +483,52 @@ pub fn InflateStream(comptime ReaderType: type) type {
         }
 
         fn decode(self: *Self, h: *Huffman) !u16 {
-            // Fast path, read some bits and hope they're prefixes of some code
-            const prefix = try self.peekBits(PREFIX_LUT_BITS);
-            if (h.prefix_lut_len[prefix] != 0) {
-                self.discardBits(h.prefix_lut_len[prefix]);
-                return h.prefix_lut[prefix];
+            // Using u32 instead of u16 to reduce the number of casts needed.
+            var prefix: u32 = 0;
+
+            // Fast path, read some bits and hope they're the prefix of some code.
+            // We can't read PREFIX_LUT_BITS as we don't want to read past the
+            // deflate stream end, use an incremental approach instead.
+            var code_len = h.min_code_len;
+            while (true) {
+                _ = try self.peekBits(code_len);
```

Comment:
Could this be made faster by instead using `bits_left` and then clamping it between `min_code_len` and `PREFIX_LUT_BITS`?

Reply:
Well it's already doing something like that if I understand your idea, the upper bound is not checked as the extra bits are masked away and the bogus `peekBits` is only used to make sure we have enough bits.

This part of the decoding process is extremely hot so please experiment with it, we may squeeze some other juice out of this without bigger changes (such as introducing secondary/tertiary lookup tables for longer codes).

Reply:
Is there a way to test the performance of the implementation already set up?

Reply:
I'm using this simple snippet:

```zig
const std = @import("std");
const io = std.io;
const mem = std.mem;
const math = std.math;
const zlib = std.compress.zlib;

pub fn main() !void {
    const DATA = @embedFile("/tmp/big.z");

    var fbs = io.fixedBufferStream(DATA);
    var buf: [16384]u8 = undefined;

    var zlib_stream = try zlib.zlibStream(std.heap.page_allocator, fbs.reader());
    defer zlib_stream.deinit();

    while (true) {
        const r = try zlib_stream.read(&buf);
        if (r == 0) break;
    }
}
```

Where `z.big` is obtained with `head -c10000000 /dev/urandom | xxd | zpipe > /tmp/big.z` and [this](https://github.com/madler/zlib/blob/master/examples/zpipe.c) is zpipe.